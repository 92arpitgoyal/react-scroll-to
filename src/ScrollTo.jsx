import PropTypes from "prop-types";
import handleScroll from "./handleScroll";

/**
 * Component that uses render props to inject
 * a function that allows the consumer to scroll to a
 * position in the window.
 */
function ScrollTo(props) {
  return props.children && props.children(handleScroll);
}

ScrollTo.defaultProps = {
  children: () => {}
};

ScrollTo.propTypes = {
  children: PropTypes.func.isRequired
};

export default ScrollTo;
