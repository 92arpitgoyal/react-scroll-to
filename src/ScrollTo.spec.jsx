import React from "react";
import PropTypes from "prop-types";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import ScrollTo from "./ScrollTo";

beforeEach(() => {
  window.scroll = jest.fn();
});

describe("Test render prop.", () => {
  it("Should render the functional children.", () => {
    const wrapper = shallow(<ScrollTo>{scroll => <div>test</div>}</ScrollTo>);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("Should render nothing nothing on undefined.", () => {
    const wrapper = shallow(<ScrollTo>{undefined}</ScrollTo>);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("Should call window.scroll.", () => {
    const wrapper = shallow(
      <ScrollTo>
        {scroll => <button onClick={() => scroll(100, 200)}>test</button>}
      </ScrollTo>
    );

    const buttonEl = wrapper.find("button");
    buttonEl.simulate("click");

    expect(window.scroll).toHaveBeenCalledTimes(1);
    expect(window.scroll.mock.calls[0]).toEqual([100, 200]);
  });

  it("Should call window.scroll with default x,y when no arguments are provided.", () => {
    const wrapper = shallow(
      <ScrollTo>
        {scroll => <button onClick={() => scroll()}>test</button>}
      </ScrollTo>
    );

    const buttonEl = wrapper.find("button");
    buttonEl.simulate("click");

    expect(window.scroll).toHaveBeenCalledTimes(1);
    expect(window.scroll.mock.calls[0]).toEqual([0, 0]);
  });

  it("Should pass correct context to children.", () => {
    const wrapper = shallow(
      <ScrollTo>
      {scroll => <div>Test</div>}
      </ScrollTo>
    );

    expect(wrapper.instance().getChildContext()).toMatchSnapshot();
  });

  it("Should add scroll area.", () => {
    const wrapper = shallow(
      <ScrollTo>
        {scroll => <div>Test</div>}
      </ScrollTo>
    );

    const childContext = wrapper.instance().getChildContext()
    childContext.addScrollArea("foo");

    expect(wrapper.instance().scrollArea).toMatchSnapshot("foo");
  });

  it("Should remove scroll area.", () => {
    const wrapper = shallow(
      <ScrollTo>
        {scroll => <div>Test</div>}
      </ScrollTo>
    );
    const childContext = wrapper.instance().getChildContext()
    childContext.addScrollArea("foo");

    childContext.removeScrollArea("foo");

    expect(wrapper.instance().scrollArea).toEqual([]);
  });

  it("Should update scroll position of ScrollArea's if present, rather than window", () => {
    const mockNode = {
      scrollLeft: 0,
      scrollTop: 0
    };
    const wrapper = shallow(
      <ScrollTo>
        {scroll => <button onClick={() => scroll(100, 200)}>test</button>}
      </ScrollTo>
    );
    const childContext = wrapper.instance().getChildContext()
    childContext.addScrollArea(mockNode);

    const buttonEl = wrapper.find("button");
    buttonEl.simulate("click");

    expect(mockNode).toMatchSnapshot();
  });
});
