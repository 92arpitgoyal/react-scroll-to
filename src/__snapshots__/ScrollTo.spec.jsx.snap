// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render prop. Should render nothing nothing on undefined. 1`] = `""`;

exports[`Test render prop. Should render the functional children. 1`] = `
<div>
  test
</div>
`;
